io/src/main/scala/fs2/io/net/Socket.scala

```diff
@@ -69,3 +76,139 @@ trait Socket[F[_]] {
     */
   def writes: Pipe[F, Byte, INothing]
 }
+
+object Socket {
+  private[net] def forAsync[F[_]: Async](
+      ch: AsynchronousSocketChannel
+  ): Resource[F, Socket[F]] =
+    Resource.make {
+      (Semaphore[F](1), Semaphore[F](1)).mapN { (readSemaphore, writeSemaphore) =>
+        new AsyncSocket[F](ch, readSemaphore, writeSemaphore)
+      }
+    }(_ => Async[F].delay(if (ch.isOpen) ch.close else ()))
+
+  private final class AsyncSocket[F[_]](
+      ch: AsynchronousSocketChannel,
+      readSemaphore: Semaphore[F],
+      writeSemaphore: Semaphore[F]
+  )(implicit F: Async[F])
+      extends Socket[F] {
+    private[this] var readBuffer: ByteBuffer = ByteBuffer.allocateDirect(8192)
+
+    private def getBufferForRead(size: Int): F[ByteBuffer] = F.delay {
+      if (readBuffer.capacity() < size)
+        readBuffer = ByteBuffer.allocateDirect(size)
+      else
+        (readBuffer: Buffer).limit(size)
+      readBuffer
+    }
+
+    /** Performs a single channel read operation in to the supplied buffer. */
+    private def readChunk(buffer: ByteBuffer): F[Int] =
+      F.async_[Int] { cb =>
+        ch.read(
+          buffer,
+          null,
+          new CompletionHandler[Integer, AnyRef] {
+            def completed(result: Integer, attachment: AnyRef): Unit =
+              cb(Right(result))
+            def failed(err: Throwable, attachment: AnyRef): Unit =
+              cb(Left(err))
+          }
+        )
```

Comment:
We can extract this `CompletionHandler` out into a class that takes the `cb`. 

```Scala
private class CallBackHandler[Res, Attach](cb: Either[Res, Throwable] => Unit) 
  extends CompletionHandler[Res, Attach] {
    def completed(result: Res, attachment: Attach): Unit =  cb(Right(result))
    def failed(err: Throwable, attachment: Attach): Unit =  cb(Left(err)) 
}
```

With it, we can replace these lines with: 
```suggestion
      F.async_[Int] { cb => 
        ch.read(buffer, null, new CallBackHandler[Integer, AnyRef](cb))
```